### PR TITLE
Add DataTransformationFilter Module

### PR DESCRIPTION
**Pull request by CytheY**
Implemented a DataTransformation Module which enables the modification of data when fetched from a data source through a TransformationFilter. 

Custom user java script code gets executed on the ObjectNode returned by the SourceAdapter.

The filter needs to be put into the processor chain like this:

{
      "name" : "TransformationFilter",
      "arguments" : {
      	"transformationFunction": "function transform(doc){ if(doc != null){ doc.extension = \"This is an extension\"; } return doc; };"
      }
}

### CODE REVIEW

File: server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.jvalue.commons.utils.Log;
+
+import javax.script.ScriptException;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.net.URL;
```
Comment (andreas-bauer) [inline server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java]: Please remove unused import statements.

File: server/src/main/java/org/jvalue/ods/transformation/TransformationFunction.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package org.jvalue.ods.transformation;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Objects;
+import org.jvalue.ods.api.views.DataView;
```
Comment (andreas-bauer) [inline server/src/main/java/org/jvalue/ods/transformation/TransformationFunction.java]: Unused import statement.

File: server/src/test/java/org/jvalue/ods/processor/filter/TransformationFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.jvalue.ods.api.sources.DataSourceMetaData;
+import org.jvalue.ods.transformation.DataTransformationManager;
+import org.jvalue.ods.transformation.ExecutionEngine;
+import org.jvalue.ods.transformation.NashornExecutionEngine;
+
+@RunWith(JMockit.class)
+public final class TransformationFilterTest {
+
+	@Mocked
+	private MetricRegistry registry;
```
Comment (andreas-bauer) [inline server/src/test/java/org/jvalue/ods/processor/filter/TransformationFilterTest.java]: `registry` is never used.

File: server/src/test/java/org/jvalue/ods/processor/filter/TransformationFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		innerNode.put("temp", 0);
+		baseNode.set("main", innerNode);
+		baseNode.put("key2", "value1");
+
+		this.executionEngine = new NashornExecutionEngine();
+		this.transformationManager = new DataTransformationManager(executionEngine);
+	}
+
+
+	private DataSource createDataSource(String jsonPointer) {
```
Comment (andreas-bauer) [inline server/src/test/java/org/jvalue/ods/processor/filter/TransformationFilterTest.java]: Could you move helper methods after the last test method.

Comment (andreas-bauer) [reply server/src/test/java/org/jvalue/ods/processor/filter/TransformationFilterTest.java]: And also the final strings with example data.

File: server/src/main/resources/js/JsonUtil.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+transformationWrapper(mvrzesmadgnxpxqhwllv);
```
Comment (andreas-bauer) [inline server/src/main/resources/js/JsonUtil.js]: This file is a Nashorn wrapper, so it should be named `NashornWrapper.js`

File: server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	@Override
+	public ObjectNode execute(ObjectNode data, TransformationFunction transformationFunction)
+		throws ScriptException, IOException {
+		initNashornSandbox();
+		try {
+			//append custom transformation function to wrapper script
+			String script = wrapperScript + transformationFunction.getTransformFunction();
+
+			//execute script
+			nashornSandbox.inject(JSON_STRING_VAR, data.toString());
```
Comment (andreas-bauer) [inline server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java]: I don't like this inject method.
It is possible to call a function from the script an pass an argument.
```
Invocable invocable = nashornSandbox.getSandboxedInvocable();
nashornSandbox.eval(script);
result = (String) invocable.invokeFunction("transformationWrapper", data.toString());
```

Comment (CytheY) [reply server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java]: I found that the sandbox has a getSandboxedInvocable();, using this now

File: server/src/main/resources/js/SampleWeatherData [hunk 7]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (andreas-bauer) [inline server/src/main/resources/js/SampleWeatherData]: The file should have a JSON file extension.

File: server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java (hunk 6 above)
Comment (andreas-bauer) [reply server/src/main/java/org/jvalue/ods/transformation/NashornExecutionEngine.java]: sounds good

**Pull request merged**